app/opensearch/src/main/java/de/komoot/photon/opensearch/OpenSearchStructuredSearchHandler.java

```diff
@@ -0,0 +1,85 @@
+package de.komoot.photon.opensearch;
+
+import de.komoot.photon.*;
+import de.komoot.photon.searcher.StructuredSearchHandler;
+import de.komoot.photon.searcher.PhotonResult;
+import de.komoot.photon.query.StructuredPhotonRequest;
+import org.opensearch.client.json.jackson.JacksonJsonpGenerator;
+import org.opensearch.client.opensearch.OpenSearchClient;
+import org.opensearch.client.opensearch._types.SearchType;
+import org.opensearch.client.opensearch._types.query_dsl.Query;
+import org.opensearch.client.opensearch.core.SearchResponse;
+import org.opensearch.client.opensearch.core.explain.ExplanationDetail;
+import org.opensearch.common.util.iterable.Iterables;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.io.IOException;
+
+/**
+ * Execute a structured forward lookup on an Elasticsearch database.
+ */
+public class OpenSearchStructuredSearchHandler implements StructuredSearchHandler {
+    private final OpenSearchClient client;
+    private final String[] supportedLanguages;
+    private final String queryTimeout;
+
+    public OpenSearchStructuredSearchHandler(OpenSearchClient client, String[] languages, int queryTimeoutSec) {
+        this.client = client;
+        this.supportedLanguages = languages;
+        queryTimeout = queryTimeoutSec + "s";
+    }
+
+    @Override
+    public List<PhotonResult> search(StructuredPhotonRequest photonRequest) {
+        var queryBuilder = buildQuery(photonRequest, false);
+
+        // for the case of deduplication we need a bit more results, #300
+        int limit = photonRequest.getLimit();
+        int extLimit = limit > 1 ? (int) Math.round(photonRequest.getLimit() * 1.5) : 1;
+
+        var results = sendQuery(queryBuilder.buildQuery(), extLimit);
+
+        if (results.hits().total().value() == 0) {
+            results = sendQuery(buildQuery(photonRequest, true).buildQuery(), extLimit);
+
+            if (results.hits().total().value() == 0 && photonRequest.hasStreet()) {
+                var street = photonRequest.getStreet();
```

Comment:
this turned out to be easier than having an unmaintainable complex query.
Of course this could also be done on  the client side...